main_test.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"syscall"
+	"testing"
+	"time"
+
+	models "github.com/firecracker-microvm/firecracker-go-sdk/client/models"
+)
+
+var kernelImagePath = "./vmlinux"
+var rootDrivePath = "/dev/null"
+
+func init() {
+	if v := os.Getenv("KERNELIMAGE"); len(v) != 0 {
+		kernelImagePath = v
+	}
+
+	if v := os.Getenv("ROOTDRIVE"); len(v) != 0 {
+		rootDrivePath = v
+	}
+}
+
+func TestFireCTL(t *testing.T) {
+	const firectlName = "./firectl"
+	if _, err := os.Stat(firectlName); os.IsNotExist(err) {
+		t.Skip("Missing firectl binary")
+	}
+
+	if _, err := os.Stat(kernelImagePath); os.IsNotExist(err) {
+		t.Skip("Missing kernel image")
+	}
+
+	if _, err := os.Stat(rootDrivePath); os.IsNotExist(err) {
+		t.Skip("Missing root drive")
+	}
+
+	ctx := context.Background()
+	ctlCtx, cancelFn := context.WithTimeout(ctx, 500*time.Millisecond)
+	defer cancelFn()
+
+	const socketpath = "./integration_test.sock"
+	firectlArgs := []string{
+		"-s",
+		socketpath,
+		"--kernel",
+		kernelImagePath,
+		"--root-drive",
+		rootDrivePath,
+	}
+	cmd := exec.CommandContext(ctlCtx, firectlName, firectlArgs...)
+	if err := cmd.Start(); err != nil {
+		t.Fatalf("Failed to run command: %v", err)
+	}
+
+	defer func() {
+		if err := os.Remove(socketpath); err != nil {
+			t.Log("Failed to remove socket path file")
+		}
+
+		if err := cmd.Process.Signal(syscall.SIGQUIT); err != nil {
```

Comment:
Why `SIGQUIT` and not `SIGTERM`?  `SIGQUIT` usually [requests a core dump](https://www.gnu.org/software/libc/manual/html_node/Termination-Signals.html).

Reply:
`SIGQUIT` also cleans up resources, reading up on it a little. My interpretation of `SIGTERM` is if we don't handle that signal, then Firecracker's process may leak.

Reply:
Not quite.  Both `SIGQUIT` and `SIGTERM` are catch-able signals (see [the man page](http://man7.org/linux/man-pages/man7/signal.7.html)), they just have different conventions and different defaults around them.  If you look at the table describing the signals, you'll see that `SIGQUIT` has a default action of [`Core`](http://man7.org/linux/man-pages/man5/core.5.html), which terminates the process and creates a core dump file, while `SIGTERM` has a default action of `Term` which just terminates the process (cleanly).  The convention is that `SIGTERM` is the signal used to ask a process to gracefully exit (along with `SIGINT`, which is the signal generated when someone presses ctrl-c in a typical terminal) and many programs install handlers specifically for `SIGTERM` so they can gracefully exit.  `SIGQUIT` is less frequently used with a non-default handler, so you may have an unclean exit and a generated core dump file with the contents of memory.